### PR TITLE
Reduce memory utilization of Route/VirtualHost.

### PR DESCRIPTION
**Pull request by yurykats**
For optional config sub-messages, use unique_ptr members and a default singleton implementation.

This is a step towards more memory efficient config data structures. 

This PR covers RetryPolicy, HedgePolicy and  InternalRedirectPolicy.

Issue #24154.

Signed-off-by: Yury Kats <ykats@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/router/config_impl.h [hunk 1]
```diff
@@ -387,6 +387,7 @@ class RetryPolicyImpl : public RetryPolicy {
   ProtobufMessage::ValidationVisitor* validation_visitor_{};
   std::vector<Upstream::RetryOptionsPredicateConstSharedPtr> retry_options_predicates_;
 };
+using DefaultRetryPolicy = ConstSingleton<RetryPolicyImpl>;
```
Comment (adisuissa) [inline source/common/router/config_impl.h:390]: nit: I think this and the other aliases can be moved into the `RouteEntryImplBase` class, to decrease namespace pollution.

File: source/common/router/config_impl.h [hunk 2]
```diff
@@ -241,10 +241,10 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
   bool includeAttemptCountInRequest() const override { return include_attempt_count_in_request_; }
   bool includeAttemptCountInResponse() const override { return include_attempt_count_in_response_; }
   const std::vector<ShadowPolicyPtr>& shadowPolicies() const { return shadow_policies_; }
-  const absl::optional<envoy::config::route::v3::RetryPolicy>& retryPolicy() const {
+  const std::unique_ptr<envoy::config::route::v3::RetryPolicy>& retryPolicy() const {
```
[the remark this answers is not in the document]
Comment (yurykats) [reply source/common/router/config_impl.h]: Great suggestion, thanks. Done

File: source/common/router/config_impl.h [hunk 3]
```diff
@@ -241,10 +241,10 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
...
     return retry_policy_;
   }
-  const absl::optional<envoy::config::route::v3::HedgePolicy>& hedgePolicy() const {
+  const std::unique_ptr<envoy::config::route::v3::HedgePolicy>& hedgePolicy() const {
```
[the remark this answers is not in the document]
Comment (yurykats) [reply source/common/router/config_impl.h]: Done

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -1123,50 +1123,52 @@ RouteEntryImplBase::parseOpaqueConfig(const envoy::config::route::v3::Route& rou
   return ret;
 }
 
-HedgePolicyImpl RouteEntryImplBase::buildHedgePolicy(
-    const absl::optional<envoy::config::route::v3::HedgePolicy>& vhost_hedge_policy,
+std::unique_ptr<HedgePolicyImpl> RouteEntryImplBase::buildHedgePolicy(
```
[the remark this answers is not in the document]
Comment (yurykats) [reply source/common/router/config_impl.cc:1126]: I personally think this is readable. I tried to see if I can be complient with the current code style in this file, but it's already inconsistent, with fair share of aliased and non-aliased types. If there are no strong objections, I'd leave this be.

File: source/common/router/config_impl.h [hunk 5]
```diff
@@ -241,11 +241,17 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
   bool includeAttemptCountInRequest() const override { return include_attempt_count_in_request_; }
   bool includeAttemptCountInResponse() const override { return include_attempt_count_in_response_; }
   const std::vector<ShadowPolicyPtr>& shadowPolicies() const { return shadow_policies_; }
-  const absl::optional<envoy::config::route::v3::RetryPolicy>& retryPolicy() const {
-    return retry_policy_;
+  const OptRef<envoy::config::route::v3::RetryPolicy> retryPolicy() const {
+    if (retry_policy_ != nullptr) {
+      return *retry_policy_;
+    }
+    return ansl::nullopt;
```
Comment (jmarantz) [inline source/common/router/config_impl.h:248]: typo?

Comment (yurykats) [reply source/common/router/config_impl.h]: Duh, thanks, fixed

File: source/common/router/config_impl.h [hunk 6]
```diff
@@ -241,11 +241,17 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
   bool includeAttemptCountInRequest() const override { return include_attempt_count_in_request_; }
   bool includeAttemptCountInResponse() const override { return include_attempt_count_in_response_; }
   const std::vector<ShadowPolicyPtr>& shadowPolicies() const { return shadow_policies_; }
-  const absl::optional<envoy::config::route::v3::RetryPolicy>& retryPolicy() const {
-    return retry_policy_;
+  const OptRef<envoy::config::route::v3::RetryPolicy> retryPolicy() const {
```
Comment (adisuissa) [inline source/common/router/config_impl.h:244]: Should the return type be `OptRef<const T>` (here and for `hedgePolicy()`)?

**Review by adisuissa — COMMENTED** [review]
Thanks! Left one small question, but otherwise LGTM.

File: source/common/router/config_impl.h (hunk 6 above)
Comment (yurykats) [reply source/common/router/config_impl.h]: Done

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Comment by yurykats** [comment]
@zuercher please take a look, thanks!

**Review by zuercher — APPROVED** [review]

**Pull request merged**